Complex variants might use a more declarative style

alalazo opened the pull request:
fixes #6314 

To sum up the issue: `multi-valued` variants with default set to `None` were not parsable from the command line, as `None` was converted to the empty string.

This PR solves the problem by sanitizing all the packages against having `None` as a default value, and allows for the more declarative syntax discussed in the issue.

The implementation is based on the possibility that the `variant` directive now has to accept as the `values` argument an object that could prescribe its own default value or other settings. A declarative style can be maintained using a [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) to set these embedded arguments in such an object.

A few packages have been modified to show how the new (optional) syntax looks like.

**Example output with this PR**
```console
# The default value can now be specified explicitly
$ spack spec openmpi fabrics=auto
Input spec
--------------------------------
openmpi fabrics=auto

Concretized
--------------------------------
openmpi@3.1.2%gcc@8.0.1~cuda+cxx_exceptions fabrics=auto ~java~legacylaunchers~memchecker~pmi schedulers=auto ~sqlite3~thread_multiple+vt arch=linux-ubuntu18.04-x86_64
...

# We can easily write a custom error message
$ spack spec openmpi fabrics=auto,psm
Input spec
--------------------------------
openmpi fabrics=auto,psm

Concretized
--------------------------------
==> Error: the value 'auto' is mutually exclusive with any of the other values [openmpi, variant 'fabrics']

$ spack spec mvapich2 process_managers=slurm,gforker
Input spec
--------------------------------
mvapich2 process_managers=gforker,slurm

Concretized
--------------------------------
==> Error: 'slurm' or 'auto' cannot be activated along with other process managers [mvapich2, variant 'process_managers']

# Finally let's suppose I left `default=None` in a new package, here's what happens
# (This will fail the package_sanity unit test on Travis)
$ spack spec mvapich2 
Input spec
--------------------------------
mvapich2

Concretized
--------------------------------
==> Error: either a default was not explicitly set, or 'None' was used [mvapich2, variant 'ch3_rank_bits']
```

scheibelp commented:
I can look at this tomorrow (10/11)

ax3l commented on var/spack/repos/builtin/packages/mvapich2/package.py:
```diff
@@ -113,8 +108,7 @@ class Mvapich2(AutotoolsPackage):
         description='List of the ROMIO file systems to activate',
-        values=('lustre', 'gpfs', 'nfs', 'ufs'),
-        multi=True
+        values=auto_or_any_combination_of('lustre', 'gpfs', 'nfs', 'ufs'),
```
auto is auto-detection of the configure step in many cases. this is different to `none` where it's explicitly none of the options.

ax3l replied:
admittedly, in most of my own (CMake) scripts where auto-detection is an option, I make them explicitly binary in packaging recipes for full control.

alalazo replied:
It's meant for auto-detection (as @ax3l said) in contrast to `none` which should be used when no feature is activated. The idea is that in some options (like this one) if you don't set any of the values the build-system will choose something automatically for you. Another example of this kind is the variant `fabrics` for some `mpi` implementations (as you can't have an `mpi` without *any* fabric support).

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -465,23 +465,50 @@ def variant(
+    # The object defining variant values might supply its own defaults for
+    # all the other arguments. Ensure we have no conflicting definitions
+    # in place.
+    for argument in ('default', 'multi', 'validator'):
```
> Associating these functions under some common namespace like multi would also help to clarify things.

I agree with changing the error message but not with this point. In this PR I am using a variant function for a multi-valued variant but the implementation is such that we could use it for any kind of variant in the future, should the need arise. I'd like to maintain this flexibility, as it comes for free. Would the message:
```
cannot specify "{0}" when passing a variant object to the argument "values" of a variant directive [variant "{name}", package "{pkg.name}"]
```
be fine with you?

> I think a user would be confused by this message, since the variant functions implicitly define these values. 

Just to clarify a thing, in case it's needed. This error is not meant for a normal user (a guy that just want to install software with Spack) as a package failing in this loop won't pass the `test_package_sanity` unit test in Travis. It will be seen only by packagers.

scheibelp replied:
> In this PR I am using a variant function for a multi-valued variant but the implementation is such that we could use it for any kind of variant in the future

On one hand I get what you mean. On the other hand in the future this could be refactored such that it receives objects that set `multi`, `default`, and `validator` so you can trace back to the class generating the conflict. To be clear I don't think you would have to handle that now - a TODO would be sufficient.

> a package failing in this loop won't pass the test_package_sanity unit test in Travis. It will be seen only by packagers.

That's who I had in mind when I complained about the error message. When possible I want to reduce the level of sophistication required for writing packages. Also if a package writer hits this issue I'm pretty sure they'd end up opening an issue, and it took me a few minutes from looking at the code to understand what was actually going on.

alalazo replied:
Minor modification to your suggestion:
`Remove specification of {0} argument: it is handled by an attribute of the "values" argument`

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
+        with_non_feature_values(single_value)
+
+
+def any_combination_of(*values):
```
To me `any_combination_of` includes the empty set, which is represented by `none`. Can we just improve the docstring of this function in case? I'd like to keep the names as brief as possible (without sacrificing clarity).

scheibelp replied:
Perhaps this function (or both functions) could have an `_allow_empty_set` option which defaults to `True` and adds the `"none"` option.

`auto_or_any_combination_of` does not include "none", so I think including that parameter would be a good compromise between keeping clear but concise function names and naming consistency.

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
         spec.variants.substitute(new_variant)
 
 
+class DisjointSetsOfValues(Sequence):
```
> Minimally, this would require some updates to the documentation (I need to track down precisely where) - **although I assume you were holding off on that until a conversation was had about the PR**

:+1: 

> While the invocations read off the tongue more smoothly, IMO it is not inherently more comprehensible than a functional approach.

I agree with that for simple variants. For variants with a more complicated logic I think this approach reads better than a functional one and permits to set defaults in a central place (the factory functions like `any_combination_of`). Imo it's also in line with the declarative nature of directives.

> (i.e. as they appear in the context of their prose - what does "with" mean for a builder object? I still have to read the function description or look at the code).

Maybe I don't get your point above, but I don't see where this isn't clear:
```
values=any_combination_of('foo', 'bar').with_default('bar')
```
in the context of the variant directive.

scheibelp replied:
> Maybe I don't get your point above, but I don't see where this isn't clear:
> `values=any_combination_of('foo', 'bar').with_default('bar')`
> in the context of the variant directive.

Reading this is fine (in this case), but when a user looks up how to write it, they may be momentarily confused by the paradigm shift. Most questions of "how do I set this or that" boil down to passing the write arguments to the write functions, but in this case it's now a matter of invoking a series of functions; those unfamiliar with it might find it slightly jarring (and to be clear I only think this is worth mentioning because this is exposed to package writers).

alalazo commented on var/spack/repos/builtin/packages/scr/package.py:
```diff
@@ -64,7 +64,7 @@ class Scr(CMakePackage):
             description='Location for SCR to find its system config file. '
             'May be either absolute or relative to the install prefix')
-    variant('copy_config', default=None,
+    variant('copy_config', default='none',
```
Thanks, I just didn't see that!

alalazo replied:
This should be fixed by the latest commit.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -465,23 +465,54 @@ def variant(
+    if default is None or default == '':
+        def _raise_default_not_set(pkg):
+            msg = "the default value in variant '{0}' from package" \
+                  " '{1}' needs to be set explicitly"
```
Where can I edit the docs to address these points:

> * How they code the possibility of choosing no value
> * How they specify this on the command line (e.g. when the default is not "none")

?

scheibelp replied:
@alalazo I missed this question:

My first thought was that there would be a section on variants in the package writing guide (https://spack.readthedocs.io/en/latest/packaging_guide.html#packaging-guide) and that would be the place to put it, but apparently there isn't any such section.

IMO this PR would ideally add a variants section to the packaging guide, and it would mention these conventions. If preferred, I can add a commit to this PR which adds that documentation.

alalazo replied:
I should have time to add docs + push forward this PR over the week-end. Will ping you when done.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -465,23 +465,54 @@ def variant(
         else:
             values = lambda x: True
 
-    if default is None:
```
> I looked through various packages to see if they omitted a default for boolean variants, and I couldn't find an example. It is however a behavior that was supported before by this logic. I didn't see explicit mention of this being supported in the docs, e.g. https://spack.readthedocs.io/en/latest/packaging_guide.html, so it may not be a problem to remove it.

That's what I did too before removing these two lines. In any case, to have `default=False` for a boolean variant you had to specify:
```python
variant('foo', values=(True, False))  # default=False is inferred
```
which is way less clear to me than saying:
```python
variant('foo', default=False)  # The fact that this is a boolean variant is inferred
```

> I'd like to be more-thorough about checking that though (I suppose the unit tests ought to have enforced it if it were a true feature).

This is enforced in the `test_package_sanity` unit test.

alalazo replied:
I'll mark this conversation as resolved, given the "thumbs-up"

alalazo commented on lib/spack/spack/test/variant.py:
```diff
@@ -711,3 +714,10 @@ def test_str(self):
         assert str(c) == ' feebar=foo foo=bar,baz foobar=fee +shared'
+
+
+def test_disjoint_set_initialization_errors():
```
The packages that have been modified to use `any_combination_of` etc.  test this implicitly in the `test_package_sanity` unit test. Do you prefer to have an explicit unit test here just for `disjoint_sets`?

scheibelp replied:
I'm satisfied with what you mention (so I don't require it), but IMO it would still be worthwhile to explicitly test the simple object here. There could be issues with `DisjointSet` and additionally issues with how `DisjointSet` is used by the package object so having the separate tests would help there.

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
+    def __contains__(self, item):
+        return item in itertools.chain.from_iterable(self.sets)
+
+    def __getitem__(self, idx):
```
> I am certainly biased, but I think I've made a stronger case when we've discussed it before than just stating my opinion, and I also did that here (in a shorter format). 

So did I, I think (the reference to our previous discussion is https://github.com/spack/spack/pull/4307#discussion_r131808515). Briefly my points are:
1. `collections.abc` is  a standard library module, not some exotic third-party package.
2. Deriving from `Sequence` to code an object that mimics a `tuple` is an accepted idiom in the Python community (as it is deriving from `MutableMapping` to mimic a `dict`).
3. It's not always true that less functionality means less complexity. Probably there are more Python users that would be comfortable with the concept of `tuple` than with that of an "iterable container" (i.e. an object that implements `__iter__` and `__contains__`).

Sorry if I insist on this, but implementing just `__iter__` and `__contains__` feels like introducing a bad practice to me. Also, the changes to implement just those two methods will give more or less the same line count as it is now - so no added benefit on that front.

That said, given that my primary interest here is to fix a bug + extend the functionality we provide to packagers, if for you having `Sequence` is enough to prevent this PR from being merged I'll change the implementation on your request. In that case can you please suggest how to reword:
https://github.com/spack/spack/blob/62d821d8431f5430484a95927cd5c0a882f808ff/lib/spack/spack/directives.py#L462

as we would break the promise to accept a tuple, and somebody extending the `variant` directive might get hit by this discrepancy in the future.

scheibelp replied:
> `collections.abc` is a standard library module, not some exotic third-party package.

None of my complaints have been about `collections.abc` being third-party or obscure. Regardless of how core it is to the language, I still think it increases the complexity.

> Deriving from `Sequence` to code an object that mimics a `tuple` is an accepted idiom in the Python community

I'm not sure how to assess the value of this: how much of the Python community accepts it? Why do they accept it? Presumably there are downsides, for one this isn't interchangeable with a tuple because it doesn't implement `__hash__`. For future reference I prefer either a link to a document I can read, or an explanation of the reasons why the Python community would prefer this.

> It's not always true that less functionality means less complexity. Probably there are more Python users that would be comfortable with the concept of tuple than with that of an "iterable container" (i.e. an object that implements __iter__ and __contains__)

This highlights a disconnect between the internals of the function (and what it is capable of handling) and the documentation that should be provided to users. I agree that it is likely best to tell users "use a tuple" because that works and is what they are familiar with. From the perspective of those editing `Variant`, I expect `__iter__` to be understandable. More on this with the reword suggestion below.

Also, I'm not sure why you need to implement `__contains__` separately: that often comes "for free" with `__iter__`.

> can you please suggest how to reword:

'values' can be any object with an idempotent `__iter__` method, or a callable. As far as the user-facing aspect of the docstring goes, this is likely to be too much detail. I'd keep the docstring the same (since tuples *are* acceptable), and add a `#`-style comment below.

I'd summarize by saying that I'm still not convinced, *BUT* this class is small and self-contained enough (unlike `RepoPath` in #4307) that I'm considering letting it go. I think that some of your motivation comes from merging usage concerns and maintenance concerns, and I think at least some of this is better addressed with documentation. Minimally, I request that you add a comment near the class declaration which mentions that you are extending `Sequence` to give this object the appearance of a tuple.

When core devs put effort into fitting the smallest reasonable interface, it saves me headaches worrying about whether this will set a precedent for other Spack coders to start extending container classes (again, I still think it's not a good choice for `RepoPath`).

ax3l commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,133 @@ Go cannot be used to fetch a particular commit or branch, it always
+            extra_args.append('--enable-static-exec')
+
+As explained in :ref:`basic-variants` the constraint ``+shared`` means
+that the boolean variant is set True, while ``~shared`` means it is set
```
maybe ``True``?
```suggestion
that the boolean variant is set ``True``, while ``~shared`` means it is set
```

ax3l commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,133 @@ Go cannot be used to fetch a particular commit or branch, it always
+
+As explained in :ref:`basic-variants` the constraint ``+shared`` means
+that the boolean variant is set True, while ``~shared`` means it is set
+to False.
```
maybe ``False``?
```suggestion
to ``False``.
```

ax3l commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,133 @@ Go cannot be used to fetch a particular commit or branch, it always
+    class Gcc(AutotoolsPackage):
+        ...
+        variant(
+            'languages', default='c,c++,fortran',
```
hm, I somehow have the feeling the empty-set issue already arises here and is not addressed below yet

alalazo replied:
I chose this example because it doesn't make sense to me to have no languages selected. On top of that you can't select `none` as it's not there. Does this seem reasonable?

ax3l replied:
Ah yes, of course!

alalazo commented:
@scheibelp @ax3l I should have addressed all your comments about the code Let me know if that's not the case and I'll react asap. 

On the other hand, I think it's likely that docs will require a few more iterations. As soon as they are in a good enough state, and if you think it's fine to proceed like this, I'd propose to merge the this PR and do the final polishing / refinement of the docs in a separate PR.

scheibelp commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,133 @@ Go cannot be used to fetch a particular commit or branch, it always
+            description='Enable dataspaces and/or flexpath staging transports'
+        )
+
+and employs a fluent-style API to improve the readability of these complex cases
```
I don't think it's helpful to mention the usage of a fluent style API (I don't think users should have to be familiar with the general concept of a fluent API to use this): examples of how to set defaults etc. are best.

alalazo replied:
I added another small example, and rephrased the sentence.

scheibelp commented:
@alalazo feel free to remove the `don't-merge-yet` tag when you think things are ready.

alalazo commented on lib/spack/spack/test/build_systems.py:
```diff
@@ -123,8 +123,11 @@ def test_with_or_without(self):
         s.concretize()
```
Good catch. This is supposed to be allowed. I'll fix this and try to simplify the way in which we mark names not to be used in `with_or_without` etc.

alalazo replied:
Should be fixed in 3cc558f

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -598,6 +605,164 @@ def substitute_abstract_variants(spec):
+    """Allows combinations from one of many mutually exclusive sets.
+
+    The value ``('none',)`` is reserved to denote the empty set
+    and therefore no other set can contain the item ``'none'``.
```
> It is a strange effect that I just noticed that when 'none' is allowed, it doesn't appear in values: that's worth mentioning.

Along with the bug you found above, this is another effect I didn't notice in my previous implementation (and for which there was no test). To explain my previous choice: I tried to implement the variant object in such a way that I didn't have to modify `AutotoolsPackage`. 

Since now a modification to that class seems necessary to fix the 2 issues you raised, I went with a variation of your first suggestion:
https://github.com/spack/spack/blob/3cc558f9d237437bd725ed7ae9ea5bd12bd5df3f/lib/spack/spack/build_systems/autotools.py#L362-L374

by allowing a variant object to expose an attribute called `feature_values`. This restores the expected semantic of a tuple, brings back the possibility to ask for `foo=none` even when `none` is not the default and shows `none` in the list of values.

alalazo commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,133 @@ Go cannot be used to fetch a particular commit or branch, it always
+        ...
+        variant(
+            'threads', default='none', description='Multithreading support',
+            values=('pthreads', 'openmp', 'none'), multi=False
```
Should be fixed in ab6693c

alalazo commented:
@scheibelp Ready for another round!

alalazo commented:
@scheibelp

alalazo commented:
@scheibelp

scheibelp commented:
> can we push this forward?

I can look at this by 11/27

alalazo commented:
Thanks @scheibelp !

scheibelp commented on lib/spack/spack/build_systems/autotools.py:
```diff
@@ -359,9 +359,18 @@ def _activate_or_not(
             options = [(name, condition in spec)]
         else:
             condition = '{name}={value}'
+            # If the "values" of a variant are a fully-fledged object with
```
I think a simpler explanation would be:

> "feature_values" is used to track values which correspond to features which can be enabled or disabled as understood by the package's build system. It excludes values which have special meanings and do not correspond to features (e.g. "none")

And actually, given the importance, IMO `feature_values` should be a documented property of `DisjointSets`

Also, I'd prefer to avoid referencing a PR in an explanatory comment: IMO it's fine for unit tests but it's worth striving for self-contained clarity in the core.

alalazo replied:
Fixed.

alalazo commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1077,6 +1077,174 @@ Go cannot be used to fetch a particular commit or branch, it always
+
+  .. code-block:: python
+
+    class Mvapich2(AutotoolsPackage):
```
Agree with almost everything of the above. The only thing I didn't implement is a convenience function that automatically defines `auto` as an allowed value on a disjoint set. 

I think `auto` is sometimes necessary (and we are flexible enough when we need it), but it's better not to encourage packagers towards using it as it potentially maps the same hash to different binaries  - e.g. the automatic discovery of a feature might depend on things that are not under the direct control of Spack. Does that make sense to you?

scheibelp replied:
That sounds OK, but in that case I think this example and the `Mvapich2` package description should use the `with_non_feature_values` function. Otherwise a user would trip over `auto` if the `Mvapich2` package definition is updated to use the `with_or_without` Autotools convenience functionality.

alalazo replied:
Makes sense to me. Done.

alalazo commented:
@scheibelp

alalazo commented:
@scheibelp

scheibelp commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -785,3 +789,49 @@ def test_spec_flags_maintain_order(self):
+        # We can't have conflicting definitions for arguments
+        fn = variant(
+            'foo', values=spack.variant.any_combination_of('fee', 'foom'),
+            default='bar'
```
@alalazo when I first skimmed this I missed that this is not failing because `'bar'` doesn't exist but rather because `variant.any_combination_of` is made responsible for specifying the default.

We've talked before about the error message and I like it, but when I examine this particular test I get the feeling there is a conflict of sorts between the `DisjointSets` constructor logic (using the fluent interface) and the typical `variant()` declaration logic.

I think it would be totally fine to throw this error if the user did something like:

```
fn = variant(
  'foo',
  values=spack.variant.any_combination_of('fee', 'foom')
    .with_default('fee'),
  default='fee'
)
```

But if the user didn't explicitly specify `with_default`, then ideally setting the `default=` argument within the `variant` would override the implicit default chosen by the `DisjointSets` constructor.

I think this is more of an issue for `'default'` rather than for `'multi'` or `'validator'`

Does that complaint make sense? I think that minimally it is worth adding a TODO for it.

alalazo replied:
TODO added

alalazo commented:
@scheibelp Let me know if anything else is needed here.

alalazo commented:
@scheibelp

scheibelp commented:
Thanks!

alalazo commented:
@scheibelp Thanks to you for the reviews! The PR improved a lot from its original version.
